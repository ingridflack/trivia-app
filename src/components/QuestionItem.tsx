import {
  FormControl,
  FormErrorMessage,
  Grid,
  Radio,
  RadioGroup,
  Text,
} from "@chakra-ui/react";
import DOMPurify from "dompurify";
import { Control, Controller, FieldErrors } from "react-hook-form";
import { AnswerFormValues } from "../pages/Trivia";
import { TriviaQuestion } from "../services/triviaService";

interface QuestionItemProps {
  question?: TriviaQuestion;
  control: Control<AnswerFormValues>;
  errors: FieldErrors<AnswerFormValues>;
}

export default function QuestionItem({
  question,
  errors,
  control,
}: QuestionItemProps) {
  if (!question) return null;

  return (
    <>
      <Text
        fontSize={{ base: "xl", md: "x-large" }}
        fontWeight="bold"
        lineHeight="1.2"
        width="100%"
        dangerouslySetInnerHTML={{
          __html: DOMPurify.sanitize(question.question),
        }}
      />

      <FormControl isInvalid={!!errors.answer} flex="1">
        <Controller
          control={control}
          name="answer"
          render={({ field }) => (
            <RadioGroup {...field}>
              <Grid
                templateColumns={{ base: "1fr", md: "repeat(2, 1fr)" }}
                gap="20px"
                marginTop="30px"
              >
                {question.answers.map((option) => (
                  <Radio key={option} colorScheme="purple" value={option}>
                    <Text
                      dangerouslySetInnerHTML={{
                        __html: DOMPurify.sanitize(option),
                      }}
                    />
                  </Radio>
                ))}
              </Grid>
            </RadioGroup>
          )}
        />
        <FormErrorMessage>
          {errors.answer && errors.answer.message}
        </FormErrorMessage>
      </FormControl>
    </>
  );
}
