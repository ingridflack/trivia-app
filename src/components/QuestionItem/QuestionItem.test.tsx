import { afterEach, describe, expect, it } from "vitest";
import QuestionItem from ".";
import { cleanup, fireEvent, render, renderHook } from "@testing-library/react";
import { useForm } from "react-hook-form";

const { result } = renderHook(() => useForm());

describe("QuestionItem", () => {
  afterEach(() => {
    cleanup();
  });

  it("should match snapshot", () => {
    const { container } = render(
      <QuestionItem
        question={{
          _id: "1",
          difficulty: "easy",
          question: "What is the capital of France?",
          answers: ["Paris", "London", "Berlin", "Madrid"],
        }}
        // @ts-expect-error- ignore the error for the purpose of the test
        control={result.current.control}
        errors={{}}
      />
    );
    expect(container).toMatchSnapshot();
  });

  it("should match the snapshot when question is undefined", () => {
    const { container } = render(
      <QuestionItem
        question={undefined}
        // @ts-expect-error- ignore the error for the purpose of the test
        control={result.current.control}
        errors={{}}
      />
    );

    expect(container).toMatchSnapshot();
  });

  it("should render the question and answer options", () => {
    const { getByText, getAllByRole } = render(
      <QuestionItem
        question={{
          _id: "1",
          difficulty: "easy",
          question: "What is the capital of France?",
          answers: ["Paris", "London", "Berlin", "Madrid"],
        }}
        // @ts-expect-error- ignore the error for the purpose of the test
        control={result.current.control}
        errors={{}}
      />
    );

    expect(getByText(/What is the capital of France?/i)).toBeTruthy();
    expect(getAllByRole("radio")).toHaveLength(4);
  });

  it("should select the answer when the user clicks on an option", () => {
    const { getAllByRole } = render(
      <QuestionItem
        question={{
          _id: "1",
          difficulty: "easy",
          question: "What is the capital of France?",
          answers: ["Paris", "London", "Berlin", "Madrid"],
        }}
        // @ts-expect-error- ignore the error for the purpose of the test
        control={result.current.control}
        errors={{}}
      />
    );

    const options = getAllByRole("radio");
    const [paris, london, berlin, madrid] = options.map(
      (option) => option.parentElement
    );

    if (!paris || !london || !berlin || !madrid) {
      throw new Error("Option not found");
    }

    expect(paris.hasAttribute("data-checked")).toBeFalsy();
    expect(london.hasAttribute("data-checked")).toBeFalsy();
    expect(berlin.hasAttribute("data-checked")).toBeFalsy();
    expect(madrid.hasAttribute("data-checked")).toBeFalsy();

    fireEvent.click(paris);

    expect(paris.hasAttribute("data-checked")).toBeTruthy();
    expect(london.hasAttribute("data-checked")).toBeFalsy();
    expect(berlin.hasAttribute("data-checked")).toBeFalsy();
    expect(madrid.hasAttribute("data-checked")).toBeFalsy();
  });

  it("should render the error message when the user submits without selecting an answer", () => {
    const { getByText } = render(
      <QuestionItem
        question={{
          _id: "1",
          difficulty: "easy",
          question: "What is the capital of France?",
          answers: ["Paris", "London", "Berlin", "Madrid"],
        }}
        // @ts-expect-error- ignore the error for the purpose of the test
        control={result.current.control}
        errors={{
          answer: { type: "required", message: "This field is required" },
        }}
      />
    );

    expect(getByText(/This field is required/i)).toBeTruthy();
  });
});
