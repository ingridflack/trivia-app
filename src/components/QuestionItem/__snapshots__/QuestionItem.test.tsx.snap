// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`QuestionItem > should match snapshot 1`] = `
<div>
  <p
    class="chakra-text css-1wwbc3v"
  >
    What is the capital of France?
  </p>
  <div
    class="chakra-form-control css-1rr4qq7"
    role="group"
  >
    <div
      class="chakra-radio-group css-0"
      role="radiogroup"
    >
      <div
        class="css-1sehaq3"
      >
        <label
          class="chakra-radio css-1nqw95c"
        >
          <input
            class="chakra-radio__input"
            id="radio-:r2:"
            name="answer"
            style="border: 0px; clip: rect(0px, 0px, 0px, 0px); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
            type="radio"
            value="Paris"
          />
          <span
            aria-hidden="true"
            class="chakra-radio__control css-ssalds"
          />
          <span
            class="chakra-radio__label css-1y8kf23"
          >
            <p
              class="chakra-text css-0"
            >
              Paris
            </p>
          </span>
        </label>
        <label
          class="chakra-radio css-1nqw95c"
        >
          <input
            class="chakra-radio__input"
            id="radio-:r3:"
            name="answer"
            style="border: 0px; clip: rect(0px, 0px, 0px, 0px); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
            type="radio"
            value="London"
          />
          <span
            aria-hidden="true"
            class="chakra-radio__control css-ssalds"
          />
          <span
            class="chakra-radio__label css-1y8kf23"
          >
            <p
              class="chakra-text css-0"
            >
              London
            </p>
          </span>
        </label>
        <label
          class="chakra-radio css-1nqw95c"
        >
          <input
            class="chakra-radio__input"
            id="radio-:r4:"
            name="answer"
            style="border: 0px; clip: rect(0px, 0px, 0px, 0px); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
            type="radio"
            value="Berlin"
          />
          <span
            aria-hidden="true"
            class="chakra-radio__control css-ssalds"
          />
          <span
            class="chakra-radio__label css-1y8kf23"
          >
            <p
              class="chakra-text css-0"
            >
              Berlin
            </p>
          </span>
        </label>
        <label
          class="chakra-radio css-1nqw95c"
        >
          <input
            class="chakra-radio__input"
            id="radio-:r5:"
            name="answer"
            style="border: 0px; clip: rect(0px, 0px, 0px, 0px); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
            type="radio"
            value="Madrid"
          />
          <span
            aria-hidden="true"
            class="chakra-radio__control css-ssalds"
          />
          <span
            class="chakra-radio__label css-1y8kf23"
          >
            <p
              class="chakra-text css-0"
            >
              Madrid
            </p>
          </span>
        </label>
      </div>
    </div>
  </div>
</div>
`;

exports[`QuestionItem > should match the snapshot when question is undefined 1`] = `<div />`;
